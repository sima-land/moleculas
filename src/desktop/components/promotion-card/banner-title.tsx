import React from 'react';
import { PromotionType } from './types';
import { Timer } from '@sima-land/ui-nucleons/timer';
import { formatISO } from 'date-fns';
import styles from './banner-title.module.scss';

export interface BannerTitleProps {
  promotionType: PromotionType;
  volumeDiscount?: number;
  dueDate?: Date;
}

/**
 * Заголовок баннера.
 * @param props Свойства.
 * @return Элемент.
 */
export const BannerTitle = ({ promotionType, volumeDiscount, dueDate }: BannerTitleProps) => {
  let content: React.ReactNode = null;

  switch (promotionType) {
    case '3-by-2':
      content = '1+1=3';
      break;
    case 'special':
      if (dueDate) {
        content = <CustomTimer dueDate={dueDate} />;
      }
      break;
    case 'volume-discount':
      if (volumeDiscount) {
        content = (
          <div className={styles.discount} data-testid='promotion-card:banner-discount'>
            скидки до
            <div className={styles.percent}>{volumeDiscount}%</div>
          </div>
        );
      }
      break;
  }

  return <>{content}</>;
};

/**
 * Таймер.
 * @param props Свойства.
 * @return Элемент.
 */
export const CustomTimer = ({ dueDate }: { dueDate: Date }) => (
  <Timer
    date={formatISO(dueDate)}
    timeout={1000 * 60}
    format={({ days, hours, minutes }: any) => (
      <div className={styles.timer} data-testid='promotion-card:banner-timer'>
        <TimerPart label='дней' value={Math.min(99, days)} />
        <TimerDivider />
        <TimerPart label='часов' value={hours % 24} />
        <TimerDivider />
        <TimerPart label='минут' value={minutes % 60} />
      </div>
    )}
  />
);

/**
 * Часть таймера.
 * @param props Свойства.
 * @return Элемент.
 */
const TimerPart = ({ label, value }: { label: string; value: number }) => (
  <div className={styles.part}>
    <div>{`${value}`.padStart(2, '0')}</div>
    <div className={styles.label}>{label}</div>
  </div>
);

/**
 * Разделитель частей таймера.
 * @return Элемент.
 */
const TimerDivider = () => <div className={styles.divider}>:</div>;
