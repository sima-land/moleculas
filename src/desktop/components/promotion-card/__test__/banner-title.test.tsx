import { mount } from 'enzyme';
import { BannerTitle, BannerTitleProps } from '../banner-title';
import { addDays } from 'date-fns';

describe('<BannerTitle />', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date('2021-01-01T09:00:00'));
  });

  it('should renders correctly', () => {
    const variants: BannerTitleProps[] = [
      // 3 by 2
      {
        promotionType: '3-by-2',
      },

      // special
      {
        promotionType: 'special',
        dueDate: undefined,
      },
      {
        promotionType: 'special',
        dueDate: addDays(new Date(), 9),
      },

      // discount
      {
        promotionType: 'volume-discount',
        volumeDiscount: undefined,
      },
      {
        promotionType: 'volume-discount',
        volumeDiscount: 45,
      },
    ];

    variants.forEach(variant => {
      const wrapper = mount(<BannerTitle {...variant} />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
