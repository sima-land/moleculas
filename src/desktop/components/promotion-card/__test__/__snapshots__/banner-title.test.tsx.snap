// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BannerTitle /> should renders correctly 1`] = `
<div>
  1+1=3
</div>
`;

exports[`<BannerTitle /> should renders correctly 2`] = `<div />`;

exports[`<BannerTitle /> should renders correctly 3`] = `
<div>
  <div
    class="timer"
    data-testid="promotion-card:banner-timer"
  >
    <div
      class="part"
    >
      <div>
        09
      </div>
      <div
        class="label"
      >
        дней
      </div>
    </div>
    <div
      class="divider"
    >
      :
    </div>
    <div
      class="part"
    >
      <div>
        00
      </div>
      <div
        class="label"
      >
        часов
      </div>
    </div>
    <div
      class="divider"
    >
      :
    </div>
    <div
      class="part"
    >
      <div>
        00
      </div>
      <div
        class="label"
      >
        минут
      </div>
    </div>
  </div>
</div>
`;

exports[`<BannerTitle /> should renders correctly 4`] = `<div />`;

exports[`<BannerTitle /> should renders correctly 5`] = `
<div>
  <div
    class="discount"
    data-testid="promotion-card:banner-discount"
  >
    скидки до
    <div
      class="percent"
    >
      45
      %
    </div>
  </div>
</div>
`;
